### PR TITLE
Update code that is pending deprecation

### PR DESCRIPTION
**Pull request by timmysilv**
While much of the code still works (sometimes), this is to future-proof us when Qiskit changes the details of their API.

### CODE REVIEW

File: doc/devices/ibmq.rst [hunk 1]
```diff
@@ -34,6 +34,11 @@ You may also directly pass your IBM Q API token to the device:
 
     dev = qml.device('qiskit.ibmq', wires=2, backend='ibmq_qasm_simulator', ibmqx_token="XXX")
 
+You may also save your token as an environment variable by running the following in a terminal:
+
+.. code::
+
+    export IBMQX_TOKEN=<my_token>
```
Comment (rmoyard) [inline doc/devices/ibmq.rst:41]: :+1:

File: doc/requirements.txt [hunk 2]
```diff
@@ -10,6 +10,7 @@ pygments==2.7.4
 pygments-github-lexers==0.0.5
 qiskit==0.42.1
 qiskit-ibm-runtime==0.9.3
```
Comment (rmoyard) [inline doc/requirements.txt:12]: Probably needs a bump, same for qiskit

File: pennylane_qiskit/aer.py [hunk 3]
```diff
@@ -31,16 +31,16 @@ class AerDevice(QiskitDevice):
...
 
     Args:
         wires (int or Iterable[Number, str]]): Number of subsystems represented by the device,
             or iterable that contains unique labels for the subsystems as numbers (i.e., ``[-1, 0, 2]``)
             or strings (``['ancilla', 'q1', 'q2']``).
         backend (str): the desired backend
         method (str): The desired simulation method. A list of supported simulation
-            methods can be returned using ``qiskit.Aer.available_methods()``, or by referring
-            to the ``AerSimulator`` `documentation <https://qiskit.org/documentation/stubs/qiskit.providers.aer.AerSimulator.html>`__.
+            methods can be returned using ``qiskit_aer.AerSimulator().available_methods()``, or by referring
```
Comment (rmoyard) [inline pennylane_qiskit/aer.py:42]: They did not make things simpler...

File: pennylane_qiskit/ibmq.py [hunk 4]
```diff
@@ -82,14 +85,20 @@ def batch_execute(self, circuits):  # pragma: no cover
     def _track_run(self):  # pragma: no cover
         """Provide runtime information."""
 
+        expected_keys = {"created", "running", "finished"}
         time_per_step = self._current_job.time_per_step()
+        if not set(time_per_step).issuperset(expected_keys):
+            self._current_job.wait_for_final_state(timeout=60)  # seconds
```
Comment (rmoyard) [inline pennylane_qiskit/ibmq.py:91]: I think here it could be much more than 60 seconds.

File: pennylane_qiskit/ibmq.py [hunk 5]
```diff
@@ -82,14 +85,20 @@ def batch_execute(self, circuits):  # pragma: no cover
     def _track_run(self):  # pragma: no cover
         """Provide runtime information."""
 
+        expected_keys = {"created", "running", "finished"}
```
Comment (rmoyard) [inline pennylane_qiskit/ibmq.py:88]: Does Qiskit provide those keys somewhere?

File: pennylane_qiskit/runtime_devices.py [hunk 6]
```diff
@@ -55,6 +56,7 @@ class IBMQCircuitRunnerDevice(IBMQDevice):
     def __init__(self, wires, provider=None, backend="ibmq_qasm_simulator", shots=1024, **kwargs):
         self.kwargs = kwargs
         super().__init__(wires=wires, provider=provider, backend=backend, shots=shots, **kwargs)
+        self.runtime_service = QiskitRuntimeService(channel="ibm_quantum")
```
Comment (rmoyard) [inline pennylane_qiskit/runtime_devices.py:59]: What are the other possibilities for channel?

File: pennylane_qiskit/runtime_devices.py [hunk 7]
```diff
@@ -199,14 +200,13 @@ def generate_samples(self, circuit_id=None):
...
-        counts = self._current_job.get("quasi_dists")[circuit_id]
+        counts = self._current_job.quasi_dists[circuit_id]
         keys = list(counts.keys())
 
-        number_of_states = 2 ** len(keys[0])
-
-        # Convert state to int
-        for i, elem in enumerate(keys):
-            keys[i] = int(elem, 2)
+        if isinstance(keys[0], str):
```
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane_qiskit/runtime_devices.py:206]: I think they changed the behavior there. Are there other available options than quasi_dists?

File: pennylane_qiskit/runtime_devices.py [hunk 8]
```diff
@@ -162,11 +165,9 @@ def batch_execute(self, circuits):
             program_inputs[kwarg] = self.kwargs.get(kwarg)
 
         # Specify the backend.
-        options = {"backend_name": self.backend.name()}
+        options = {"backend": self.backend.name}
         # Send circuits to the cloud for execution by the sampler program.
-        job = self.provider.runtime.run(
-            program_id="sampler", options=options, inputs=program_inputs
-        )
+        job = self.runtime_service.run(program_id="sampler", options=options, inputs=program_inputs)
```
Comment (rmoyard) [inline pennylane_qiskit/runtime_devices.py:170]: Why is the sampler still available as a runtime program? I think we have to make it a primitive (in the future)

File: requirements.txt [hunk 9]
```diff
@@ -26,9 +26,9 @@ pycparser==2.21
 python-constraint==1.4.0
 python-dateutil==2.8.2
 qiskit==0.42.1
-qiskit-aer==0.10.1
+qiskit-aer==0.12.0
 qiskit-ibm-runtime==0.9.3
```
Comment (rmoyard) [inline requirements.txt:30]: We can probably bump qiskit and provider.

**Review by rmoyard — COMMENTED** [review]
Thanks @timmysilv, great job :1st_place_medal: I have left a couple of comments. Sorry that you had to rework the test IBMQ test suite.

File: pennylane_qiskit/aer.py (hunk 3 above)
Comment (timmysilv) [reply pennylane_qiskit/aer.py:42]: nope : (

File: pennylane_qiskit/ibmq.py (hunk 5 above)
Comment (timmysilv) [reply pennylane_qiskit/ibmq.py:88]: I dug deep into API docs to find [this](https://cloud.ibm.com/apidocs/quantum-computing#get-job-metrics-jid-response). If you expand the `timestamps` drop-down in the `Response` category, you see these keys. Of course, you'd only check that if you found that `_time_per_step` gets its metadata from the `timestamps` endpoint by digging through source code (ctrl-F for "timestamps" in [this file](https://qiskit.org/ecosystem/ibm-provider/_modules/qiskit_ibm_provider/job/ibm_circuit_job.html#IBMCircuitJob.time_per_step))

File: pennylane_qiskit/runtime_devices.py (hunk 6 above)
Comment (timmysilv) [reply pennylane_qiskit/runtime_devices.py:59]: per [the parameters in the docs](https://qiskit.org/documentation/partners/qiskit_ibm_runtime/stubs/qiskit_ibm_runtime.QiskitRuntimeService.html), it's this or `ibm_cloud`. Everything I saw uses `ibm_quantum` so I am sticking with that

File: pennylane_qiskit/runtime_devices.py (hunk 8 above)
Comment (timmysilv) [reply pennylane_qiskit/runtime_devices.py:170]: my guess is that "primitives" are still just runtime programs and they won't explicitly remove support for this. I did it to reduce the amount of change in this PR, because a primitive does not map 1:1 to a device, but here we call the sampler a "device". I'll make a story for this change, but I'm not sure how urgent it is.

File: requirements.txt (hunk 9 above)
Comment (timmysilv) [reply requirements.txt:30]: already at latest

File: pennylane_qiskit/runtime_devices.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane_qiskit/runtime_devices.py:206]: `self._current_job` is now a [SamplerResult](https://qiskit.org/documentation/stubs/qiskit.primitives.SamplerResult.html), so this is all we've got. The docs don't seem to address this exact concern

[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane_qiskit/runtime_devices.py:206]: I found the most minimal solution! They _look_ like dicts, but they are typed objects. I inspected one instance of one and got this:

```pycon
>>> from qiskit.result import QuasiDistribution
>>> d = QuasiDistribution({"00": 0.94, "01": 0.06})
>>> d, d.binary_probabilities()
({0: 0.94, 1: 0.06}, {'00': 0.94, '01': 0.06})
```

tl;dr `binary_probabilities` will help guarantee we get what we want. Updating now

[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane_qiskit/runtime_devices.py]: That's a great find!

File: pennylane_qiskit/ibmq.py (hunk 5 above)
Comment (rmoyard) [reply pennylane_qiskit/ibmq.py:99]: Ok let's keep as it is then
